src/Microsoft.ML.OnnxConverter/SaveOnnxCommand.cs

```diff
@@ -322,6 +352,30 @@ private void Run(IChannel ch)
                     nameof(Arguments.LoadPredictor), "We were explicitly told to load the predictor but one was not present.");
             }
 
+            // Convert back to values the KeyDataViewType "pass-through" columns
+            // (i.e those that remained untouched by the model). This is done to enable NimbusML to get these values
+            // as described in https://github.com/dotnet/machinelearning/pull/4841
+
+            var inputKeyDataViewTypeColumnsNames = new HashSet<string>();
+            foreach (var col in source.Schema)
+                if (col.IsHidden == false && HasKeyValues(col))
+                    inputKeyDataViewTypeColumnsNames.Add(col.Name);
+
+            var passThroughColumnNames = new HashSet<string>();
+            var onlyDistinctColumnNames = end.Schema.Select(c => c.Name).Distinct(); //  only check column names that appear once in the output schema
+            foreach (var col in end.Schema)
+                if (col.IsHidden == false && onlyDistinctColumnNames.Contains(col.Name) && HasKeyValues(col))
+                    passThroughColumnNames.Add(col.Name);
+
+            passThroughColumnNames.IntersectWith(inputKeyDataViewTypeColumnsNames); // Only count those columns that were in the input of the pipeline
+
+            foreach (var name in passThroughColumnNames)
```

Comment:
>passThroughColumnNames [](start = 33, length = 22)

Here is another edge case: If the input has a categorical column, say `A`, and the pipeline has a `CopyTransform` that copies `A` into a column named `B`. 
The way it is written now, we will not apply key-to-value on `B`. Should we?
On one hand, we should, because `A` and `B` should be copies of each other, so it does not make sense if the ONNX model has different types and values for them... On the other hand, I don't see an easy rule that would help decide when we do need to apply key-to-value to `B` (when it is the output of Copy), versus when we don't (for example, when it is the output of ToKey).

Reply:
Similar issue with Concat: say the input has two categorical columns `A` and `B`, and the pipeline has a concat transform that concatenates them to produce a column `C`...

---
In reply to: [383580868](https://github.com/dotnet/machinelearning/pull/4841#discussion_r383580868) [](ancestors = 383580868)

Reply:
Ideally I think we should apply KeyToValueTransformer to B. But, as you say, I can't think of an easy way to decide when to do this when it's the output of Copy. So I wouldn't know if we should handle this particular edge case.

Thoughts, @ganik , @harishsk ?

Reply:
agree, that is a bit strange side-effect. Perhaps in next iterations we can get more details on how AutoML prefers this. Maybe we wont need to add KeyToValue for passthrough columns at all and let them surface as keys. For now we can document this.